PlayList.java

```diff
@@ -71,34 +109,73 @@ public int indexOf(String title) {
      *  is full, does nothing and returns false. Otherwise, inserts the track and
      *  returns true. */
     public boolean add(int i, Track track) {
-        //// replace the following statement with your code
-        return false;
+        if(i >= maxSize || i < 0 )
+        {
+            return false;
+        }
+        if(size -1 < i)
+        {
+            tracks[size] = track;
+            size++;
+            return true;
+        }
+
+        for(int n= size; n > i; n--)
+        {
+            tracks[n]= tracks[n-1];
+        }
+        tracks[i]=track;
+        size++;
+        
+        return true;
+        
     }
      
     /** Removes the track in the given index from this list.
      *  If the list is empty, or the given index is negative or too big for this list, 
      *  does nothing and returns -1. */
     public void remove(int i) {
-        //// replace this comment with your code
+
+        if(i<maxSize || size !=0 || i>0)
+        {
+            for(int j = i; j < size-1 ; j++)
+            {
+                tracks[j]= tracks[j+1];
+            }
+            tracks[size-1] = null;
+            size--;
+        }
+             
+       
     }
 
     /** Removes the first track that has the given title from this list.
      *  If such a track is not found, or the list is empty, or the given index
      *  is negative or too big for this list, does nothing. */
     public void remove(String title) {
-        //// replace this comment with your code
+        int i = indexOf(title);
+        remove(i);
+        size--;
     }
 
     /** Removes the first track from this list. If the list is empty, does nothing. */
     public void removeFirst() {
-        //// replace this comment with your code
+        remove(0);
+        size--;
     }
     
     /** Adds all the tracks in the other list to the end of this list. 
      *  If the total size of both lists is too large, does nothing. */
     //// An elegant and terribly inefficient implementation.
      public void add(PlayList other) {
-        //// replace this comment with your code
+        if(this.size + other.getSize() <= maxSize)
+        {
+            for(int i=0; i<other.getSize();i++)
+            {
+                this.add(other.getTrack(i));
+            }
+            size+= other.size;
```

Comment:
Each `add` call will increase the size, by this logic you'll cause the size to be increased twice than the other playlist length.
```suggestion
            for(int i=0; i<other.getSize();i++)
            {
                this.add(other.getTrack(i));
            }
```